### PR TITLE
Implement OPSAW and SASAW

### PR DESCRIPTION
**Pull request by frostedoyster**

### CODE REVIEW

File: mops/src/sasaw/cpu.tpp [hunk 1]
```diff
@@ -18,7 +19,65 @@ void mops::sparse_accumulation_scatter_add_with_weights(
...
+
+    // For each index in the first dimension of the outputs,
+    // get what indices in the inputs should write to it
+    std::vector<std::vector<size_t>> write_list = get_write_list(indices_output_1);
+    std::fill(o_ptr, o_ptr+output.shape[0]*output_shift_first_dim, static_cast<scalar_t>(0.0));
+
+    #pragma omp parallel for
+    for (size_t i = 0; i < size_output_first_dim; i++) {
+        // iterate over input indices that will write to the output index i
+        for (size_t e : write_list[i]) {
```
Comment (Luthaf) [inline mops/src/sasaw/cpu.tpp:64]: I'm a bit lost as to what `e` means here. Maybe adding a comment or using a longer name would help.

File: mops/src/sasaw/cuda.tpp [hunk 2]
```diff
@@ -4,16 +4,16 @@
...
+    Tensor<scalar_t, 3>,
+    Tensor<scalar_t, 2>,
+    Tensor<scalar_t, 2>,
+    Tensor<scalar_t, 1>,
+    Tensor<scalar_t, 3>,
+    Tensor<int, 1>,
+    Tensor<int, 1>,
+    Tensor<int, 1>,
+    Tensor<int, 1>,
+    Tensor<int, 1>
```
Comment (Luthaf) [inline mops/src/sasaw/cuda.tpp:16]: these should be `int32_t`

File: mops/benchmarks/opsaw.cpp [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    auto A = std::vector<double>(60000 * 13);
+    fill_vector_random_floats(A);
+
+    auto B = std::vector<double>(60000 * 20);
+    fill_vector_random_floats(B);
+
+    auto W = std::vector<double>(1000 * 20);
+    fill_vector_random_floats(W);
+
+    auto indices_W = std::vector<int>(60000);
```
Comment (Luthaf) [inline mops/benchmarks/opsaw.cpp:18]: `int` => `int32_t` here and in the other benchmarks

File: mops/src/opsaw/cpu.tpp [hunk 4]
```diff
@@ -30,5 +76,77 @@ void mops::outer_product_scatter_add_with_weights_vjp(
...
+        }
+
+        scalar_t* grad_a_ptr = grad_A.data;
+        scalar_t* grad_b_ptr = grad_B.data;
+        scalar_t* grad_w_ptr = grad_W.data;
+        scalar_t* grad_o_ptr = grad_output.data;
+        scalar_t* a_ptr = A.data;
+        scalar_t* b_ptr = B.data;
+        scalar_t* w_ptr = W.data;
+        int* i_ptr = indices_output.data;
```
Comment (Luthaf) [inline mops/src/opsaw/cpu.tpp:115]:
```suggestion
        const int32_t* idx_o_ptr = indices_output.data;
```

File: mops/src/sasaw/cpu.tpp [hunk 5]
```diff
@@ -18,7 +19,65 @@ void mops::sparse_accumulation_scatter_add_with_weights(
...
+    scalar_t* o_ptr = output.data;
+    scalar_t* a_ptr = A.data;
+    scalar_t* b_ptr = B.data;
+    scalar_t* w_ptr = W.data;
+    scalar_t* c_ptr = C.data;
+    int* i_ptr = indices_output_1.data;
+    int* j_ptr = indices_W_1.data;
+    int* m_1_ptr = indices_A.data;
+    int* m_2_ptr = indices_W_2.data;
+    int* m_3_ptr = indices_output_2.data;
```
Comment (Luthaf) [inline mops/src/sasaw/cpu.tpp:45]:
```suggestion
    const int32_t* idx_o1_ptr = indices_output_1.data;
    const int32_t* idx_w1_ptr = indices_W_1.data;
    const int32_t* idx_a_ptr = indices_A.data;
    const int32_t* idx_w2_ptr = indices_W_2.data;
    const int32_t* idx_o2_ptr = indices_output_2.data;
```

File: mops/src/sasaw/cpu.tpp [hunk 6]
```diff
@@ -38,5 +97,94 @@ void mops::sparse_accumulation_scatter_add_with_weights_vjp(
...
+        scalar_t* grad_w_ptr = grad_W.data;
+        scalar_t* grad_o_ptr = grad_output.data;
+        scalar_t* a_ptr = A.data;
+        scalar_t* b_ptr = B.data;
+        scalar_t* w_ptr = W.data;
+        scalar_t* c_ptr = C.data;
+        int* i_ptr = indices_output_1.data;
+        int* m_1_ptr = indices_A.data;
+        int* m_2_ptr = indices_W_2.data;
+        int* m_3_ptr = indices_output_2.data;
```
Comment (Luthaf) [inline mops/src/sasaw/cpu.tpp:144]:
```suggestion
        const int32_t* idx_o1_ptr = indices_output_1.data;
        const int32_t* idx_a_ptr = indices_A.data;
        const int32_t* idx_w_ptr = indices_W_2.data;
        const int32_t* idx_o2_ptr = indices_output_2.data;
```

File: mops/src/sasaw/cpu.tpp [hunk 7]
```diff
@@ -38,11 +38,11 @@ void mops::sparse_accumulation_scatter_add_with_weights(
...
-    int* i_ptr = indices_output_1.data;
-    int* j_ptr = indices_W_1.data;
-    int* m_1_ptr = indices_A.data;
-    int* m_2_ptr = indices_W_2.data;
-    int* m_3_ptr = indices_output_2.data;
+    const int32_t* i_ptr = indices_output_1.data;
+    const int32_t* j_ptr = indices_W_1.data;
+    const int32_t* m_1_ptr = indices_A.data;
+    const int32_t* m_2_ptr = indices_W_2.data;
+    const int32_t* m_3_ptr = indices_output_2.data;
```
Comment (Luthaf) [inline mops/src/sasaw/cpu.tpp:45]: the name change suggestion was intentional =) These should be named relative to `output_2`, not `m3`

File: mops/src/sasaw/cpu.tpp [hunk 8]
```diff
@@ -18,7 +19,66 @@ void mops::sparse_accumulation_scatter_add_with_weights(
...
+
+    // For each index in the first dimension of the outputs,
+    // get what indices in the inputs should write to it
+    std::vector<std::vector<size_t>> write_list = get_write_list(indices_output_1);
+    std::fill(o_ptr, o_ptr+output.shape[0]*output_shift_first_dim, static_cast<scalar_t>(0.0));
+
+    #pragma omp parallel for
+    for (size_t i = 0; i < size_output_first_dim; i++) {
+        // iterate over input indices that will write to the output index i
+        // these are stored as entries (variable "e") in write_list[i]
```
Comment (Luthaf) [inline mops/src/sasaw/cpu.tpp:64]:
```suggestion
        // these are stored as entries (variable "e") in write_list[i]
        // for example `o_ptr_e` is a pointer to the first element of 
        // the entry (row) in the o tensor
```

Is this correct?

File: mops/src/sasaw/cpu.tpp (hunk 7 above)
Comment (frostedoyster) [reply mops/src/sasaw/cpu.tpp:45]: Done!

File: mops/src/sasaw/cpu.tpp (hunk 8 above)
Comment (frostedoyster) [reply mops/src/sasaw/cpu.tpp:65]: Correct, I've expanded on the comment

**Review by Luthaf — APPROVED** [review]

**Pull request merged**